Create guide for Exchanges

segfaultxavi opened the pull request:
Fixes #645

rg911 reviewed (changes requested):
There are also 2 most asked points that are missing:
1. Deposit from aggregate transactions, the slightly different payload format exchanges need to handle
2. Transaction Fees on withdraw: how to setup the maxFees

segfaultxavi commented on source/guides/exchanges/exchange-integration.rst:
```diff
@@ -0,0 +1,353 @@
+Components
+==========
+
+Central wallet
```
I like to emphasize that it is "Central", because it is in the middle of everything (all money goes through it). But I'll add a note calling it also "Hot", because that is probably a more popular term.

segfaultxavi commented on source/guides/exchanges/exchange-integration.rst:
```diff
@@ -0,0 +1,353 @@
+
+The exchange owns a Symbol account where all the user's deposits and withdrawals occur. This account only has the necessary amount of XYM for daily use (withdrawals and deposits), since it is the account most exposed to attacks.
+
+To increase the account security, we highly recommend to set up the wallet using Symbol's :doc:`multisig feature <../../concepts/multisig-account>`. Withdrawals can be automated using bots to ensure wallet security with **two-factor authentication**.
```
OK

segfaultxavi commented on source/guides/exchanges/exchange-integration.rst:
```diff
@@ -0,0 +1,353 @@
+Unique User ID
+--------------
+
+Each user is identified by a unique identifier (UUID) on the exchange's database. A user will deposit to the central wallet with their UUID attached as
```
This is only a proposal. In the proposed architecture, all deposits must have a memo. I'll clarify this point.

segfaultxavi commented on source/guides/exchanges/exchange-integration.rst:
```diff
@@ -0,0 +1,353 @@
+
+   Symbol's :doc:`Transfer transactions <../../concepts/transfer-transaction>` can hold an arbitrary message up to 1023 bytes long but **the first byte is treated specially by the** :doc:`Symbol SDK <../../sdk>`.
+
+   This can be a source of confusion because the receiver of a transaction does not know if the message was generated by the Symbol SDK or otherwise (for example accessing the :doc:`REST gateway <../../api>`), so it does not know if the first byte must be treated specially or not.
```
The problem is the confusion between the sender of a tx and the receiver, right? The exchange may or may not use the SDK, but it will never know if the sender of the tx used it, therefore it is safer to always ignore the first byte.
I thought this was what we agreed when we wrote this:
https://nem-software.atlassian.net/wiki/spaces/SES/pages/600605077/Memo+restrictions

segfaultxavi commented on source/guides/exchanges/exchange-integration.rst:
```diff
@@ -0,0 +1,353 @@
+
+It is advisable to turn central and cold wallets into :doc:`multisig accounts <../../concepts/multisig-account>` to add **two-factor authentication**. The cosignatories of the multisig account become the account managers, so no transaction can be announced from the multisig account without the cosignatories' approval. Symbol's current implementation of multisig is **“M-of-N”** meaning that *M* ou ...
+
+.. caution:: Multisig accounts are a **powerful** yet **dangerous** tool. If access to some cosignatory account is lost and the minimum approval is not reached (the *M* above), access to the multisig account can be permanently lost. **Always configure multisig accounts with caution**.
```
Yes, dangerous. A regular account only has to keep one key safe. A multisig N-of-N account has to keep N keys safe, plus the danger of misconfiguration. I think the Caution box is desirable.
Also, I just copied it over from the previous writer so he must think so too :)

segfaultxavi commented on source/guides/exchanges/exchange-integration.rst:
```diff
@@ -0,0 +1,353 @@
+
+   On MAINNET, these IDs are ``0x6BED913FA20223F8`` (mosaic ID) and ``0xE74B99BA41F4AFEE`` (alias ID).
+
+   **Always treat these two IDs as equivalent.**
```
I'll add a link to the guide explaining how to resolve aliases (When it is written!). But meanwhile, these two IDs must always be treated as synonyms, right? Wasn't this one of the common problems for Exchanges?

segfaultxavi commented on source/guides/exchanges/exchange-integration.rst:
```diff
@@ -0,0 +1,353 @@
+
+The procedure, which is common in blockchains which do not support finalization, is to **wait for a few blocks** to be validated (added to the blockchain) before accepting a transaction.
+
+The amount of blocks to wait for depends on the risk one wants to accept. The recommendation for Symbol is **20 blocks** (about 10 minutes, regardless of network conditions).
```
I'll add a note.

segfaultxavi commented on source/guides/exchanges/exchange-integration.rst:
```diff
@@ -0,0 +1,353 @@
+
+A bit of context is required here. Transactions are not allowed to remain unconfirmed in the network forever, as this would pose a significant strain on the network's resources. Instead, **all transactions have a deadline**, and are automatically disposed of when the deadline arrives.
+
+Users are free to use any deadline they want for their transactions, between now and 24h into the future.
```
Changed. Thanks!

segfaultxavi commented on source/guides/exchanges/exchange-integration.rst:
```diff
@@ -0,0 +1,353 @@
+
+   To avoid any issue, the following measures **must always be enforced**:
+
+   - **Always** start messages with a byte in the 32 to 128 range (this is the standard ASCII printable range).
```
See my previous comment.

segfaultxavi commented:
@rg911 I think I have addressed all your concerns, except the ones regarding the REST API (which are tracked in https://github.com/nemtech/symbol-docs/issues/645).

Changing this guide to use REST instead of the SDK is going to be a significant effort. How about I do that in another task?
Does this guide look good to you then? [You can see the preview here](https://deploy-preview-654--nemtech.netlify.app/guides/exchanges/exchange-integration.html).

segfaultxavi commented:
@rg911 I'd like to close this one. It's been sitting here for 22 days now.
I think I have addressed all your comments and the rest will be dealt in another guide (#662).

rg911 reviewed (commented):
I would put a section for `Aggregate / standalone` transactions, a brief explanation, and the payload comparison. 

Exchanges might / will expect incoming aggregate transactions (depends on how they retrieve the tx data). An example would be Zaif is sending all outgoing transactions using aggregate, if users withdraw from Zaif to other exchanges who are not handling aggregate, might see their withdrawal tx missing at the recipient side.

segfaultxavi commented:
> I would put a section for `Aggregate / standalone` transactions, a brief explanation, and the payload comparison.

I understand the problem of monitoring agg txes, but I thought I already had addressed it:
- In the SDK guide, the sample code already uses `embedded=true` and then checks the type of the returned object (`transactionInfo instanceof AggregateTransactionInfo`)
- In the REST guide there's a whole section called **Parsing of embedded transactions** about this topic. This section has a table comparing the `TransactionMetaDTO` and `EmbeddedTransactionMetaDTO` as you requested.

What else is missing?

rg911 commented:
> > I would put a section for `Aggregate / standalone` transactions, a brief explanation, and the payload comparison.
> 
> I understand the problem of monitoring agg txes, but I thought I already had addressed it:
> 
> * In the SDK guide, the sample code already uses `embedded=true` and then checks the type of the returned object (`transactionInfo instanceof AggregateTransactionInfo`)
> * In the REST guide there's a whole section called **Parsing of embedded transactions** about this topic. This section has a table comparing the `TransactionMetaDTO` and `EmbeddedTransactionMetaDTO` as you requested.
> 
> What else is missing?

yeah, saw that, would it be better to add a `note` in this one to exchange user, e.g. to expect incoming aggregate transactions

segfaultxavi commented:
I've added the note to the #674 PR, let's continue the discussion there. Having dependent PRs is not comfortable on GitHub.
